<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<html lang="en">

<head>
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
  <title>LCOV - FastANITest.out - /home/srirampc/work/FastANI/src/cgi/include/computeCoreIdentity.hpp</title>
  <link rel="stylesheet" type="text/css" href="../../../../../../../gcov.css">
</head>

<body>

  <table width="100%" border=0 cellspacing=0 cellpadding=0>
    <tr><td class="title">LCOV - code coverage report</td></tr>
    <tr><td class="ruler"><img src="../../../../../../../glass.png" width=3 height=3 alt=""></td></tr>

    <tr>
      <td width="100%">
        <table cellpadding=1 border=0 width="100%">
          <tr>
            <td width="10%" class="headerItem">Current view:</td>
            <td width="35%" class="headerValue"><a href="../../../../../../../index.html">top level</a> - <a href="index.html">home/srirampc/work/FastANI/src/cgi/include</a> - computeCoreIdentity.hpp<span style="font-size: 80%;"> (source / <a href="computeCoreIdentity.hpp.func-sort-c.html">functions</a>)</span></td>
            <td width="5%"></td>
            <td width="15%"></td>
            <td width="10%" class="headerCovTableHead">Hit</td>
            <td width="10%" class="headerCovTableHead">Total</td>
            <td width="15%" class="headerCovTableHead">Coverage</td>
          </tr>
          <tr>
            <td class="headerItem">Test:</td>
            <td class="headerValue">FastANITest.out</td>
            <td></td>
            <td class="headerItem">Lines:</td>
            <td class="headerCovTableEntry">54</td>
            <td class="headerCovTableEntry">183</td>
            <td class="headerCovTableEntryLo">29.5 %</td>
          </tr>
          <tr>
            <td class="headerItem">Date:</td>
            <td class="headerValue">2022-09-28 16:45:31</td>
            <td></td>
            <td class="headerItem">Functions:</td>
            <td class="headerCovTableEntry">3</td>
            <td class="headerCovTableEntry">9</td>
            <td class="headerCovTableEntryLo">33.3 %</td>
          </tr>
          <tr><td><img src="../../../../../../../glass.png" width=3 height=3 alt=""></td></tr>
        </table>
      </td>
    </tr>

    <tr><td class="ruler"><img src="../../../../../../../glass.png" width=3 height=3 alt=""></td></tr>
  </table>

  <table cellpadding=0 cellspacing=0 border=0>
    <tr>
      <td><br></td>
    </tr>
    <tr>
      <td>
<pre class="sourceHeading">          Line data    Source code</pre>
<pre class="source">
<a name="1"><span class="lineNum">       1 </span>            : /**</a>
<a name="2"><span class="lineNum">       2 </span>            :  * @file    computeCoreIdentity.hpp</a>
<a name="3"><span class="lineNum">       3 </span>            :  * @author  Chirag Jain &lt;cjain7@gatech.edu&gt;</a>
<a name="4"><span class="lineNum">       4 </span>            :  */</a>
<a name="5"><span class="lineNum">       5 </span>            : </a>
<a name="6"><span class="lineNum">       6 </span>            : #ifndef CGI_IDENTITY_HPP </a>
<a name="7"><span class="lineNum">       7 </span>            : #define CGI_IDENTITY_HPP</a>
<a name="8"><span class="lineNum">       8 </span>            : </a>
<a name="9"><span class="lineNum">       9 </span>            : #include &lt;vector&gt;</a>
<a name="10"><span class="lineNum">      10 </span>            : #include &lt;algorithm&gt;</a>
<a name="11"><span class="lineNum">      11 </span>            : #include &lt;unordered_map&gt;</a>
<a name="12"><span class="lineNum">      12 </span>            : #include &lt;fstream&gt;</a>
<a name="13"><span class="lineNum">      13 </span>            : #include &lt;omp.h&gt;</a>
<a name="14"><span class="lineNum">      14 </span>            : #include &lt;zlib.h&gt;  </a>
<a name="15"><span class="lineNum">      15 </span>            : </a>
<a name="16"><span class="lineNum">      16 </span>            : //Own includes</a>
<a name="17"><span class="lineNum">      17 </span>            : #include &quot;map/include/base_types.hpp&quot;</a>
<a name="18"><span class="lineNum">      18 </span>            : #include &quot;cgi/include/cgid_types.hpp&quot;</a>
<a name="19"><span class="lineNum">      19 </span>            : </a>
<a name="20"><span class="lineNum">      20 </span>            : //External includes</a>
<a name="21"><span class="lineNum">      21 </span>            : #include &quot;common/kseq.h&quot;</a>
<a name="22"><span class="lineNum">      22 </span>            : #include &quot;common/prettyprint.hpp&quot;</a>
<a name="23"><span class="lineNum">      23 </span>            : </a>
<a name="24"><span class="lineNum">      24 </span>            : namespace cgi</a>
<a name="25"><span class="lineNum">      25 </span>            : {</a>
<a name="26"><span class="lineNum">      26 </span>            :   /**</a>
<a name="27"><span class="lineNum">      27 </span>            :    * @brief                       Use reference sketch's sequence to file (genome) mapping </a>
<a name="28"><span class="lineNum">      28 </span>            :    *                              and revise reference ids to genome id</a>
<a name="29"><span class="lineNum">      29 </span>            :    * @param[in/out] shortResults</a>
<a name="30"><span class="lineNum">      30 </span>            :    */</a>
<a name="31"><span class="lineNum">      31 </span><span class="lineCov">          7 :   void reviseRefIdToGenomeId(std::vector&lt;MappingResult_CGI&gt; &amp;shortResults, skch::Sketch &amp;refSketch)</span></a>
<a name="32"><span class="lineNum">      32 </span>            :   {</a>
<a name="33"><span class="lineNum">      33 </span><span class="lineCov">      38980 :     for(auto &amp;e : shortResults)</span></a>
<a name="34"><span class="lineNum">      34 </span>            :     {</a>
<a name="35"><span class="lineNum">      35 </span><span class="lineCov">      38973 :       auto referenceSequenceId = e.refSequenceId;</span></a>
<a name="36"><span class="lineNum">      36 </span><span class="lineCov">      38973 :       auto upperRangeIter = std::upper_bound(refSketch.sequencesByFileInfo.begin(), </span></a>
<a name="37"><span class="lineNum">      37 </span>            :           refSketch.sequencesByFileInfo.end(),</a>
<a name="38"><span class="lineNum">      38 </span>            :           referenceSequenceId);</a>
<a name="39"><span class="lineNum">      39 </span>            : </a>
<a name="40"><span class="lineNum">      40 </span><span class="lineCov">      38973 :       e.genomeId = std::distance(refSketch.sequencesByFileInfo.begin(), upperRangeIter);</span></a>
<a name="41"><span class="lineNum">      41 </span>            :     }</a>
<a name="42"><span class="lineNum">      42 </span><span class="lineCov">          7 :   }</span></a>
<a name="43"><span class="lineNum">      43 </span>            : </a>
<a name="44"><span class="lineNum">      44 </span>            :   /**</a>
<a name="45"><span class="lineNum">      45 </span>            :    * @brief                       compute genome lengths in reference and query genome set</a>
<a name="46"><span class="lineNum">      46 </span>            :    * @param[out] genomeLengths</a>
<a name="47"><span class="lineNum">      47 </span>            :    */</a>
<a name="48"><span class="lineNum">      48 </span><span class="lineNoCov">          0 :   void computeGenomeLengths(skch::Parameters &amp;parameters, std::unordered_map &lt;std::string, uint64_t&gt; &amp;genomeLengths) </span></a>
<a name="49"><span class="lineNum">      49 </span>            :   { </a>
<a name="50"><span class="lineNum">      50 </span><span class="lineNoCov">          0 :     for(auto &amp;e : parameters.querySequences)</span></a>
<a name="51"><span class="lineNum">      51 </span>            :     {</a>
<a name="52"><span class="lineNum">      52 </span>            :       //Open the file using kseq</a>
<a name="53"><span class="lineNum">      53 </span><span class="lineNoCov">          0 :       gzFile fp = gzopen(e.c_str(), &quot;r&quot;);</span></a>
<a name="54"><span class="lineNum">      54 </span><span class="lineNoCov">          0 :       kseq_t *seq = kseq_init(fp);</span></a>
<a name="55"><span class="lineNum">      55 </span><span class="lineNoCov">          0 :       int l; uint64_t genomeLen = 0;</span></a>
<a name="56"><span class="lineNum">      56 </span>            : </a>
<a name="57"><span class="lineNum">      57 </span><span class="lineNoCov">          0 :       while ((l = kseq_read(seq)) &gt;= 0) {</span></a>
<a name="58"><span class="lineNum">      58 </span><span class="lineNoCov">          0 :         if (l &gt;= parameters.minReadLength) {</span></a>
<a name="59"><span class="lineNum">      59 </span><span class="lineNoCov">          0 :           uint64_t _l_ = (((uint64_t)strlen(seq-&gt;seq.s)) / parameters.minReadLength) * parameters.minReadLength;</span></a>
<a name="60"><span class="lineNum">      60 </span><span class="lineNoCov">          0 :           genomeLen = genomeLen + _l_;</span></a>
<a name="61"><span class="lineNum">      61 </span>            :         }</a>
<a name="62"><span class="lineNum">      62 </span>            :       }</a>
<a name="63"><span class="lineNum">      63 </span>            : </a>
<a name="64"><span class="lineNum">      64 </span><span class="lineNoCov">          0 :       genomeLengths[e] = genomeLen;</span></a>
<a name="65"><span class="lineNum">      65 </span>            : </a>
<a name="66"><span class="lineNum">      66 </span><span class="lineNoCov">          0 :       kseq_destroy(seq);  </span></a>
<a name="67"><span class="lineNum">      67 </span><span class="lineNoCov">          0 :       gzclose(fp); //close the file handler </span></a>
<a name="68"><span class="lineNum">      68 </span>            :     }</a>
<a name="69"><span class="lineNum">      69 </span>            : </a>
<a name="70"><span class="lineNum">      70 </span><span class="lineNoCov">          0 :     for(auto &amp;e : parameters.refSequences)</span></a>
<a name="71"><span class="lineNum">      71 </span>            :     {</a>
<a name="72"><span class="lineNum">      72 </span><span class="lineNoCov">          0 :       if( genomeLengths.find(e) == genomeLengths.end() )</span></a>
<a name="73"><span class="lineNum">      73 </span>            :       {</a>
<a name="74"><span class="lineNum">      74 </span>            :         //Open the file using kseq</a>
<a name="75"><span class="lineNum">      75 </span><span class="lineNoCov">          0 :         gzFile fp = gzopen(e.c_str(), &quot;r&quot;);</span></a>
<a name="76"><span class="lineNum">      76 </span><span class="lineNoCov">          0 :         kseq_t *seq = kseq_init(fp);</span></a>
<a name="77"><span class="lineNum">      77 </span><span class="lineNoCov">          0 :         int l; uint64_t genomeLen = 0;</span></a>
<a name="78"><span class="lineNum">      78 </span>            : </a>
<a name="79"><span class="lineNum">      79 </span><span class="lineNoCov">          0 :       while ((l = kseq_read(seq)) &gt;= 0) {</span></a>
<a name="80"><span class="lineNum">      80 </span><span class="lineNoCov">          0 :         if (l &gt;= parameters.minReadLength) {</span></a>
<a name="81"><span class="lineNum">      81 </span><span class="lineNoCov">          0 :           uint64_t _l_ = (((uint64_t)strlen(seq-&gt;seq.s)) / parameters.minReadLength) * parameters.minReadLength;</span></a>
<a name="82"><span class="lineNum">      82 </span><span class="lineNoCov">          0 :           genomeLen = genomeLen + _l_;</span></a>
<a name="83"><span class="lineNum">      83 </span>            :         }</a>
<a name="84"><span class="lineNum">      84 </span>            :       }</a>
<a name="85"><span class="lineNum">      85 </span>            : </a>
<a name="86"><span class="lineNum">      86 </span><span class="lineNoCov">          0 :         genomeLengths[e] = genomeLen;</span></a>
<a name="87"><span class="lineNum">      87 </span>            : </a>
<a name="88"><span class="lineNum">      88 </span><span class="lineNoCov">          0 :         kseq_destroy(seq);  </span></a>
<a name="89"><span class="lineNum">      89 </span><span class="lineNoCov">          0 :         gzclose(fp); //close the file handler </span></a>
<a name="90"><span class="lineNum">      90 </span>            :       }</a>
<a name="91"><span class="lineNum">      91 </span>            :     }</a>
<a name="92"><span class="lineNum">      92 </span><span class="lineNoCov">          0 :   }</span></a>
<a name="93"><span class="lineNum">      93 </span>            : </a>
<a name="94"><span class="lineNum">      94 </span>            :   /**</a>
<a name="95"><span class="lineNum">      95 </span>            :    * @brief                             output blast tabular mappings for visualization </a>
<a name="96"><span class="lineNum">      96 </span>            :    * @param[in]   parameters            algorithm parameters</a>
<a name="97"><span class="lineNum">      97 </span>            :    * @param[in]   results               bidirectional mappings</a>
<a name="98"><span class="lineNum">      98 </span>            :    * @param[in]   mapper                mapper object used for mapping</a>
<a name="99"><span class="lineNum">      99 </span>            :    * @param[in]   refSketch             reference sketch</a>
<a name="100"><span class="lineNum">     100 </span>            :    * @param[in]   queryFileNo           query genome is parameters.querySequences[queryFileNo]</a>
<a name="101"><span class="lineNum">     101 </span>            :    * @param[in]   fileName              file name where results will be reported</a>
<a name="102"><span class="lineNum">     102 </span>            :    */</a>
<a name="103"><span class="lineNum">     103 </span><span class="lineNoCov">          0 :   void outputVisualizationFile(skch::Parameters &amp;parameters,</span></a>
<a name="104"><span class="lineNum">     104 </span>            :       std::vector&lt;MappingResult_CGI&gt; &amp;mappings_2way,</a>
<a name="105"><span class="lineNum">     105 </span>            :       skch::Map &amp;mapper,</a>
<a name="106"><span class="lineNum">     106 </span>            :       skch::Sketch &amp;refSketch,</a>
<a name="107"><span class="lineNum">     107 </span>            :       uint64_t queryFileNo,</a>
<a name="108"><span class="lineNum">     108 </span>            :       std::string &amp;fileName)</a>
<a name="109"><span class="lineNum">     109 </span>            :   {</a>
<a name="110"><span class="lineNum">     110 </span><span class="lineNoCov">          0 :     std::ofstream outstrm(fileName + &quot;.visual&quot;, std::ios::app);</span></a>
<a name="111"><span class="lineNum">     111 </span>            : </a>
<a name="112"><span class="lineNum">     112 </span>            :     //Shift offsets for converting from local (to contig) to global (to genome)</a>
<a name="113"><span class="lineNum">     113 </span><span class="lineNoCov">          0 :     std::vector &lt;skch::offset_t&gt; queryOffsetAdder (mapper.metadata.size());</span></a>
<a name="114"><span class="lineNum">     114 </span><span class="lineNoCov">          0 :     std::vector &lt;skch::offset_t&gt; refOffsetAdder (refSketch.metadata.size());</span></a>
<a name="115"><span class="lineNum">     115 </span>            : </a>
<a name="116"><span class="lineNum">     116 </span><span class="lineNoCov">          0 :     for(int i = 0; i &lt; mapper.metadata.size(); i++)</span></a>
<a name="117"><span class="lineNum">     117 </span>            :     {</a>
<a name="118"><span class="lineNum">     118 </span><span class="lineNoCov">          0 :       if(i == 0)</span></a>
<a name="119"><span class="lineNum">     119 </span><span class="lineNoCov">          0 :         queryOffsetAdder[i] = 0;</span></a>
<a name="120"><span class="lineNum">     120 </span>            :       else</a>
<a name="121"><span class="lineNum">     121 </span><span class="lineNoCov">          0 :         queryOffsetAdder[i] = queryOffsetAdder[i-1] + mapper.metadata[i-1].len;</span></a>
<a name="122"><span class="lineNum">     122 </span>            :     }</a>
<a name="123"><span class="lineNum">     123 </span>            : </a>
<a name="124"><span class="lineNum">     124 </span><span class="lineNoCov">          0 :     for(int i = 0; i &lt; refSketch.metadata.size(); i++)</span></a>
<a name="125"><span class="lineNum">     125 </span>            :     {</a>
<a name="126"><span class="lineNum">     126 </span><span class="lineNoCov">          0 :       if(i == 0)</span></a>
<a name="127"><span class="lineNum">     127 </span><span class="lineNoCov">          0 :         refOffsetAdder[i] = 0;</span></a>
<a name="128"><span class="lineNum">     128 </span>            :       else</a>
<a name="129"><span class="lineNum">     129 </span><span class="lineNoCov">          0 :         refOffsetAdder[i] = refOffsetAdder[i-1] + refSketch.metadata[i-1].len;</span></a>
<a name="130"><span class="lineNum">     130 </span>            :     }</a>
<a name="131"><span class="lineNum">     131 </span>            : </a>
<a name="132"><span class="lineNum">     132 </span>            :     //Report all mappings that contribute to core-genome identity estimate</a>
<a name="133"><span class="lineNum">     133 </span>            :     //Format the output to blast tabular way (outfmt 6)</a>
<a name="134"><span class="lineNum">     134 </span><span class="lineNoCov">          0 :     for(auto &amp;e : mappings_2way)</span></a>
<a name="135"><span class="lineNum">     135 </span>            :     {</a>
<a name="136"><span class="lineNum">     136 </span><span class="lineNoCov">          0 :       outstrm &lt;&lt; parameters.querySequences[queryFileNo]</span></a>
<a name="137"><span class="lineNum">     137 </span><span class="lineNoCov">          0 :         &lt;&lt; &quot;\t&quot; &lt;&lt; parameters.refSequences[e.genomeId]</span></a>
<a name="138"><span class="lineNum">     138 </span><span class="lineNoCov">          0 :         &lt;&lt; &quot;\t&quot; &lt;&lt; e.nucIdentity</span></a>
<a name="139"><span class="lineNum">     139 </span>            :         &lt;&lt; &quot;\t&quot; &lt;&lt; &quot;NA&quot;</a>
<a name="140"><span class="lineNum">     140 </span>            :         &lt;&lt; &quot;\t&quot; &lt;&lt; &quot;NA&quot;</a>
<a name="141"><span class="lineNum">     141 </span>            :         &lt;&lt; &quot;\t&quot; &lt;&lt; &quot;NA&quot;</a>
<a name="142"><span class="lineNum">     142 </span><span class="lineNoCov">          0 :         &lt;&lt; &quot;\t&quot; &lt;&lt; e.queryStartPos                                    + queryOffsetAdder[e.querySeqId] </span></a>
<a name="143"><span class="lineNum">     143 </span><span class="lineNoCov">          0 :         &lt;&lt; &quot;\t&quot; &lt;&lt; e.queryStartPos + parameters.minReadLength - 1     + queryOffsetAdder[e.querySeqId]</span></a>
<a name="144"><span class="lineNum">     144 </span><span class="lineNoCov">          0 :         &lt;&lt; &quot;\t&quot; &lt;&lt; e.refStartPos                                      + refOffsetAdder[e.refSequenceId]</span></a>
<a name="145"><span class="lineNum">     145 </span><span class="lineNoCov">          0 :         &lt;&lt; &quot;\t&quot; &lt;&lt; e.refStartPos + parameters.minReadLength - 1       + refOffsetAdder[e.refSequenceId]</span></a>
<a name="146"><span class="lineNum">     146 </span>            :         &lt;&lt; &quot;\t&quot; &lt;&lt; &quot;NA&quot;</a>
<a name="147"><span class="lineNum">     147 </span>            :         &lt;&lt; &quot;\t&quot; &lt;&lt; &quot;NA&quot;</a>
<a name="148"><span class="lineNum">     148 </span><span class="lineNoCov">          0 :         &lt;&lt; &quot;\n&quot;;</span></a>
<a name="149"><span class="lineNum">     149 </span>            :     }</a>
<a name="150"><span class="lineNum">     150 </span><span class="lineNoCov">          0 :   }</span></a>
<a name="151"><span class="lineNum">     151 </span>            : </a>
<a name="152"><span class="lineNum">     152 </span>            :   /**</a>
<a name="153"><span class="lineNum">     153 </span>            :    * @brief                             compute and report AAI/ANI </a>
<a name="154"><span class="lineNum">     154 </span>            :    * @param[in]   parameters            algorithm parameters</a>
<a name="155"><span class="lineNum">     155 </span>            :    * @param[in]   results               mapping results</a>
<a name="156"><span class="lineNum">     156 </span>            :    * @param[in]   mapper                mapper object used for mapping</a>
<a name="157"><span class="lineNum">     157 </span>            :    * @param[in]   refSketch             reference sketch</a>
<a name="158"><span class="lineNum">     158 </span>            :    * @param[in]   totalQueryFragments   count of total sequence fragments in query genome</a>
<a name="159"><span class="lineNum">     159 </span>            :    * @param[in]   queryFileNo           query genome is parameters.querySequences[queryFileNo]</a>
<a name="160"><span class="lineNum">     160 </span>            :    * @param[in]   fileName              file name where results will be reported</a>
<a name="161"><span class="lineNum">     161 </span>            :    * @param[out]  CGI_ResultsVector     FastANI results</a>
<a name="162"><span class="lineNum">     162 </span>            :    */</a>
<a name="163"><span class="lineNum">     163 </span><span class="lineCov">          7 :   void computeCGI(skch::Parameters &amp;parameters,</span></a>
<a name="164"><span class="lineNum">     164 </span>            :       skch::MappingResultsVector_t &amp;results,</a>
<a name="165"><span class="lineNum">     165 </span>            :       skch::Map &amp;mapper,</a>
<a name="166"><span class="lineNum">     166 </span>            :       skch::Sketch &amp;refSketch,</a>
<a name="167"><span class="lineNum">     167 </span>            :       uint64_t totalQueryFragments,</a>
<a name="168"><span class="lineNum">     168 </span>            :       uint64_t queryFileNo,</a>
<a name="169"><span class="lineNum">     169 </span>            :       std::string &amp;fileName,</a>
<a name="170"><span class="lineNum">     170 </span>            :       std::vector&lt;cgi::CGI_Results&gt; &amp;CGI_ResultsVector</a>
<a name="171"><span class="lineNum">     171 </span>            :       )</a>
<a name="172"><span class="lineNum">     172 </span>            :   {</a>
<a name="173"><span class="lineNum">     173 </span>            :     //Vector to save relevant fields from mapping results</a>
<a name="174"><span class="lineNum">     174 </span><span class="lineCov">          7 :     std::vector&lt;MappingResult_CGI&gt; shortResults;</span></a>
<a name="175"><span class="lineNum">     175 </span>            : </a>
<a name="176"><span class="lineNum">     176 </span><span class="lineCov">          7 :     shortResults.reserve(results.size());</span></a>
<a name="177"><span class="lineNum">     177 </span>            : </a>
<a name="178"><span class="lineNum">     178 </span>            :     // Note to self: For debugging any issue, it is often useful to print</a>
<a name="179"><span class="lineNum">     179 </span>            :     // shortResults, mappings_1way and mappings_2way vectors</a>
<a name="180"><span class="lineNum">     180 </span>            : </a>
<a name="181"><span class="lineNum">     181 </span>            :     ///Parse map results and save fields which we need</a>
<a name="182"><span class="lineNum">     182 </span>            :     // reference id (R), query id (Q), estimated identity (I)</a>
<a name="183"><span class="lineNum">     183 </span><span class="lineCov">      38980 :     for(auto &amp;e : results)</span></a>
<a name="184"><span class="lineNum">     184 </span>            :     {</a>
<a name="185"><span class="lineNum">     185 </span><span class="lineCov">      38973 :       shortResults.emplace_back(MappingResult_CGI{</span></a>
<a name="186"><span class="lineNum">     186 </span><span class="lineCov">      38973 :           e.refSeqId,</span></a>
<a name="187"><span class="lineNum">     187 </span>            :           0,                  //this value will be revised to genome id</a>
<a name="188"><span class="lineNum">     188 </span><span class="lineCov">      38973 :           e.querySeqId,</span></a>
<a name="189"><span class="lineNum">     189 </span><span class="lineCov">      38973 :           e.refStartPos,</span></a>
<a name="190"><span class="lineNum">     190 </span><span class="lineCov">      38973 :           e.queryStartPos,</span></a>
<a name="191"><span class="lineNum">     191 </span><span class="lineCov">      38973 :           e.refStartPos/(parameters.minReadLength - 20),</span></a>
<a name="192"><span class="lineNum">     192 </span><span class="lineCov">      38973 :           e.nucIdentity</span></a>
<a name="193"><span class="lineNum">     193 </span>            :           });</a>
<a name="194"><span class="lineNum">     194 </span>            :     }</a>
<a name="195"><span class="lineNum">     195 </span>            : </a>
<a name="196"><span class="lineNum">     196 </span>            :     /*</a>
<a name="197"><span class="lineNum">     197 </span>            :      * NOTE: We assume single file contains the sequences for single genome</a>
<a name="198"><span class="lineNum">     198 </span>            :      * We revise reference sequence id to genome (or file) id</a>
<a name="199"><span class="lineNum">     199 </span>            :      */</a>
<a name="200"><span class="lineNum">     200 </span><span class="lineCov">          7 :     reviseRefIdToGenomeId(shortResults, refSketch);</span></a>
<a name="201"><span class="lineNum">     201 </span>            : </a>
<a name="202"><span class="lineNum">     202 </span>            : </a>
<a name="203"><span class="lineNum">     203 </span>            :     //We need best reciprocal identity match for each genome, query pair</a>
<a name="204"><span class="lineNum">     204 </span><span class="lineCov">          7 :     std::vector&lt;MappingResult_CGI&gt; mappings_1way;</span></a>
<a name="205"><span class="lineNum">     205 </span><span class="lineCov">          7 :     std::vector&lt;MappingResult_CGI&gt; mappings_2way;</span></a>
<a name="206"><span class="lineNum">     206 </span>            : </a>
<a name="207"><span class="lineNum">     207 </span>            :     ///1. Code below fetches best identity match for each genome, query pair</a>
<a name="208"><span class="lineNum">     208 </span>            :     //For each query sequence, best match in the reference is preserved</a>
<a name="209"><span class="lineNum">     209 </span>            :     {</a>
<a name="210"><span class="lineNum">     210 </span>            :       //Sort the vector shortResults</a>
<a name="211"><span class="lineNum">     211 </span><span class="lineCov">          7 :       std::sort(shortResults.begin(), shortResults.end(), cmp_query_bucket);</span></a>
<a name="212"><span class="lineNum">     212 </span>            : </a>
<a name="213"><span class="lineNum">     213 </span><span class="lineCov">      38980 :       for(auto &amp;e : shortResults)</span></a>
<a name="214"><span class="lineNum">     214 </span>            :       {</a>
<a name="215"><span class="lineNum">     215 </span><span class="lineCov">      38973 :         if(mappings_1way.empty())</span></a>
<a name="216"><span class="lineNum">     216 </span><span class="lineCov">          7 :           mappings_1way.push_back(e);</span></a>
<a name="217"><span class="lineNum">     217 </span>            : </a>
<a name="218"><span class="lineNum">     218 </span><span class="lineCov">      38966 :         else if ( !(</span></a>
<a name="219"><span class="lineNum">     219 </span><span class="lineCov">      38966 :               e.genomeId == mappings_1way.back().genomeId &amp;&amp; </span></a>
<a name="220"><span class="lineNum">     220 </span><span class="lineCov">      38962 :               e.querySeqId == mappings_1way.back().querySeqId))</span></a>
<a name="221"><span class="lineNum">     221 </span>            :         {</a>
<a name="222"><span class="lineNum">     222 </span><span class="lineCov">      18762 :           mappings_1way.emplace_back(e);</span></a>
<a name="223"><span class="lineNum">     223 </span>            :         }</a>
<a name="224"><span class="lineNum">     224 </span>            :         else</a>
<a name="225"><span class="lineNum">     225 </span>            :         {</a>
<a name="226"><span class="lineNum">     226 </span><span class="lineCov">      20204 :           mappings_1way.back() = e;</span></a>
<a name="227"><span class="lineNum">     227 </span>            :         }</a>
<a name="228"><span class="lineNum">     228 </span>            :       }</a>
<a name="229"><span class="lineNum">     229 </span>            :     }</a>
<a name="230"><span class="lineNum">     230 </span>            : </a>
<a name="231"><span class="lineNum">     231 </span>            :     ///2. Now, we compute 2-way ANI</a>
<a name="232"><span class="lineNum">     232 </span>            :     //For each mapped region, and within a reference bin bucket, single best query mapping is preserved</a>
<a name="233"><span class="lineNum">     233 </span>            :     {</a>
<a name="234"><span class="lineNum">     234 </span><span class="lineCov">          7 :       std::sort(mappings_1way.begin(), mappings_1way.end(), cmp_refbin_bucket);</span></a>
<a name="235"><span class="lineNum">     235 </span>            : </a>
<a name="236"><span class="lineNum">     236 </span><span class="lineCov">      18776 :       for(auto &amp;e : mappings_1way)</span></a>
<a name="237"><span class="lineNum">     237 </span>            :       {</a>
<a name="238"><span class="lineNum">     238 </span><span class="lineCov">      18769 :         if(mappings_2way.empty())</span></a>
<a name="239"><span class="lineNum">     239 </span><span class="lineCov">          7 :           mappings_2way.push_back(e);</span></a>
<a name="240"><span class="lineNum">     240 </span>            : </a>
<a name="241"><span class="lineNum">     241 </span><span class="lineCov">      18762 :         else if ( !(</span></a>
<a name="242"><span class="lineNum">     242 </span><span class="lineCov">      18762 :               e.refSequenceId == mappings_2way.back().refSequenceId &amp;&amp; </span></a>
<a name="243"><span class="lineNum">     243 </span><span class="lineCov">      18512 :               e.mapRefPosBin == mappings_2way.back().mapRefPosBin))</span></a>
<a name="244"><span class="lineNum">     244 </span>            :         {</a>
<a name="245"><span class="lineNum">     245 </span><span class="lineCov">      15768 :           mappings_2way.emplace_back(e);</span></a>
<a name="246"><span class="lineNum">     246 </span>            :         }</a>
<a name="247"><span class="lineNum">     247 </span>            :         else</a>
<a name="248"><span class="lineNum">     248 </span>            :         {</a>
<a name="249"><span class="lineNum">     249 </span><span class="lineCov">       2994 :           mappings_2way.back() = e;</span></a>
<a name="250"><span class="lineNum">     250 </span>            :         }</a>
<a name="251"><span class="lineNum">     251 </span>            :       }</a>
<a name="252"><span class="lineNum">     252 </span>            :     }</a>
<a name="253"><span class="lineNum">     253 </span>            : </a>
<a name="254"><span class="lineNum">     254 </span>            :     {</a>
<a name="255"><span class="lineNum">     255 </span><span class="lineCov">          7 :       if(parameters.visualize)</span></a>
<a name="256"><span class="lineNum">     256 </span>            :       {</a>
<a name="257"><span class="lineNum">     257 </span><span class="lineNoCov">          0 :         outputVisualizationFile(parameters, mappings_2way, mapper, refSketch, queryFileNo, fileName);</span></a>
<a name="258"><span class="lineNum">     258 </span>            :       }</a>
<a name="259"><span class="lineNum">     259 </span>            :     }</a>
<a name="260"><span class="lineNum">     260 </span>            : </a>
<a name="261"><span class="lineNum">     261 </span>            :     //Do average for ANI/AAI computation </a>
<a name="262"><span class="lineNum">     262 </span>            :     //mappings_2way should be sorted by genomeId </a>
<a name="263"><span class="lineNum">     263 </span>            : </a>
<a name="264"><span class="lineNum">     264 </span><span class="lineCov">         18 :     for(auto it = mappings_2way.begin(); it != mappings_2way.end();)</span></a>
<a name="265"><span class="lineNum">     265 </span>            :     {</a>
<a name="266"><span class="lineNum">     266 </span><span class="lineCov">         11 :       skch::seqno_t currentGenomeId = it-&gt;genomeId;</span></a>
<a name="267"><span class="lineNum">     267 </span>            : </a>
<a name="268"><span class="lineNum">     268 </span>            :       //Bucket by genome id</a>
<a name="269"><span class="lineNum">     269 </span><span class="lineCov">         11 :       auto rangeEndIter = std::find_if(it, mappings_2way.end(), [&amp;](const MappingResult_CGI&amp; e) </span></a>
<a name="270"><span class="lineNum">     270 </span>            :           { </a>
<a name="271"><span class="lineNum">     271 </span><span class="lineCov">      15779 :             return e.genomeId != currentGenomeId; </span></a>
<a name="272"><span class="lineNum">     272 </span>            :           } );</a>
<a name="273"><span class="lineNum">     273 </span>            : </a>
<a name="274"><span class="lineNum">     274 </span><span class="lineCov">         11 :       float sumIdentity = 0.0;</span></a>
<a name="275"><span class="lineNum">     275 </span>            : </a>
<a name="276"><span class="lineNum">     276 </span><span class="lineCov">      15786 :       for(auto it2 = it; it2 != rangeEndIter; it2++)</span></a>
<a name="277"><span class="lineNum">     277 </span>            :       {</a>
<a name="278"><span class="lineNum">     278 </span><span class="lineCov">      15775 :         sumIdentity += it2-&gt;nucIdentity;</span></a>
<a name="279"><span class="lineNum">     279 </span>            :       }</a>
<a name="280"><span class="lineNum">     280 </span>            : </a>
<a name="281"><span class="lineNum">     281 </span>            :       //Save the result </a>
<a name="282"><span class="lineNum">     282 </span>            :       CGI_Results currentResult;</a>
<a name="283"><span class="lineNum">     283 </span>            : </a>
<a name="284"><span class="lineNum">     284 </span><span class="lineCov">         11 :       currentResult.qryGenomeId = queryFileNo;</span></a>
<a name="285"><span class="lineNum">     285 </span><span class="lineCov">         11 :       currentResult.refGenomeId = currentGenomeId;</span></a>
<a name="286"><span class="lineNum">     286 </span><span class="lineCov">         11 :       currentResult.countSeq = std::distance(it, rangeEndIter);</span></a>
<a name="287"><span class="lineNum">     287 </span><span class="lineCov">         11 :       currentResult.totalQueryFragments = totalQueryFragments;</span></a>
<a name="288"><span class="lineNum">     288 </span><span class="lineCov">         11 :       currentResult.identity = sumIdentity/currentResult.countSeq;</span></a>
<a name="289"><span class="lineNum">     289 </span>            : </a>
<a name="290"><span class="lineNum">     290 </span><span class="lineCov">         11 :       CGI_ResultsVector.push_back(currentResult);</span></a>
<a name="291"><span class="lineNum">     291 </span>            : </a>
<a name="292"><span class="lineNum">     292 </span>            :       //Advance the iterator it</a>
<a name="293"><span class="lineNum">     293 </span><span class="lineCov">         11 :       it = rangeEndIter;</span></a>
<a name="294"><span class="lineNum">     294 </span>            :     }</a>
<a name="295"><span class="lineNum">     295 </span><span class="lineCov">          7 :   }</span></a>
<a name="296"><span class="lineNum">     296 </span>            : </a>
<a name="297"><span class="lineNum">     297 </span>            :   /**</a>
<a name="298"><span class="lineNum">     298 </span>            :    * @brief                             output FastANI results to file</a>
<a name="299"><span class="lineNum">     299 </span>            :    * @param[in]   parameters            algorithm parameters</a>
<a name="300"><span class="lineNum">     300 </span>            :    * @param[in]   genomeLengths</a>
<a name="301"><span class="lineNum">     301 </span>            :    * @param[in]   CGI_ResultsVector     results</a>
<a name="302"><span class="lineNum">     302 </span>            :    * @param[in]   fileName              file name where results will be reported</a>
<a name="303"><span class="lineNum">     303 </span>            :    */</a>
<a name="304"><span class="lineNum">     304 </span><span class="lineNoCov">          0 :   void outputCGI(skch::Parameters &amp;parameters,</span></a>
<a name="305"><span class="lineNum">     305 </span>            :       std::unordered_map &lt;std::string, uint64_t&gt; &amp;genomeLengths,</a>
<a name="306"><span class="lineNum">     306 </span>            :       std::vector&lt;cgi::CGI_Results&gt; &amp;CGI_ResultsVector,</a>
<a name="307"><span class="lineNum">     307 </span>            :       std::string &amp;fileName)</a>
<a name="308"><span class="lineNum">     308 </span>            :   {</a>
<a name="309"><span class="lineNum">     309 </span>            :     //sort result by identity</a>
<a name="310"><span class="lineNum">     310 </span><span class="lineNoCov">          0 :     std::sort(CGI_ResultsVector.rbegin(), CGI_ResultsVector.rend());</span></a>
<a name="311"><span class="lineNum">     311 </span>            : </a>
<a name="312"><span class="lineNum">     312 </span><span class="lineNoCov">          0 :     std::ofstream outstrm(fileName);</span></a>
<a name="313"><span class="lineNum">     313 </span>            : </a>
<a name="314"><span class="lineNum">     314 </span>            :     //Report results</a>
<a name="315"><span class="lineNum">     315 </span><span class="lineNoCov">          0 :     for(auto &amp;e : CGI_ResultsVector)</span></a>
<a name="316"><span class="lineNum">     316 </span>            :     {</a>
<a name="317"><span class="lineNum">     317 </span><span class="lineNoCov">          0 :       std::string qryGenome = parameters.querySequences[e.qryGenomeId];</span></a>
<a name="318"><span class="lineNum">     318 </span><span class="lineNoCov">          0 :       std::string refGenome = parameters.refSequences[e.refGenomeId];</span></a>
<a name="319"><span class="lineNum">     319 </span>            : </a>
<a name="320"><span class="lineNum">     320 </span><span class="lineNoCov">          0 :       assert(genomeLengths.find(qryGenome) != genomeLengths.end());</span></a>
<a name="321"><span class="lineNum">     321 </span><span class="lineNoCov">          0 :       assert(genomeLengths.find(refGenome) != genomeLengths.end());</span></a>
<a name="322"><span class="lineNum">     322 </span>            : </a>
<a name="323"><span class="lineNum">     323 </span><span class="lineNoCov">          0 :       uint64_t queryGenomeLength = genomeLengths[qryGenome];</span></a>
<a name="324"><span class="lineNum">     324 </span><span class="lineNoCov">          0 :       uint64_t refGenomeLength = genomeLengths[refGenome]; </span></a>
<a name="325"><span class="lineNum">     325 </span><span class="lineNoCov">          0 :       uint64_t minGenomeLength = std::min(queryGenomeLength, refGenomeLength);</span></a>
<a name="326"><span class="lineNum">     326 </span><span class="lineNoCov">          0 :       uint64_t sharedLength = e.countSeq * parameters.minReadLength;</span></a>
<a name="327"><span class="lineNum">     327 </span>            : </a>
<a name="328"><span class="lineNum">     328 </span>            :       //Checking if shared genome is above a certain fraction of genome length</a>
<a name="329"><span class="lineNum">     329 </span><span class="lineNoCov">          0 :       if(sharedLength &gt;= minGenomeLength * parameters.minFraction)</span></a>
<a name="330"><span class="lineNum">     330 </span>            :       {</a>
<a name="331"><span class="lineNum">     331 </span>            :         outstrm &lt;&lt; qryGenome</a>
<a name="332"><span class="lineNum">     332 </span>            :           &lt;&lt; &quot;\t&quot; &lt;&lt; refGenome</a>
<a name="333"><span class="lineNum">     333 </span><span class="lineNoCov">          0 :           &lt;&lt; &quot;\t&quot; &lt;&lt; e.identity </span></a>
<a name="334"><span class="lineNum">     334 </span><span class="lineNoCov">          0 :           &lt;&lt; &quot;\t&quot; &lt;&lt; e.countSeq</span></a>
<a name="335"><span class="lineNum">     335 </span><span class="lineNoCov">          0 :           &lt;&lt; &quot;\t&quot; &lt;&lt; e.totalQueryFragments</span></a>
<a name="336"><span class="lineNum">     336 </span><span class="lineNoCov">          0 :           &lt;&lt; &quot;\n&quot;;</span></a>
<a name="337"><span class="lineNum">     337 </span>            :       }</a>
<a name="338"><span class="lineNum">     338 </span><span class="lineNoCov">          0 :     }</span></a>
<a name="339"><span class="lineNum">     339 </span>            : </a>
<a name="340"><span class="lineNum">     340 </span><span class="lineNoCov">          0 :     outstrm.close();</span></a>
<a name="341"><span class="lineNum">     341 </span><span class="lineNoCov">          0 :   }</span></a>
<a name="342"><span class="lineNum">     342 </span>            : </a>
<a name="343"><span class="lineNum">     343 </span>            :   /**</a>
<a name="344"><span class="lineNum">     344 </span>            :    * @brief                             output FastANI results as lower triangular matrix</a>
<a name="345"><span class="lineNum">     345 </span>            :    * @param[in]   parameters            algorithm parameters</a>
<a name="346"><span class="lineNum">     346 </span>            :    * @param[in]   genomeLengths</a>
<a name="347"><span class="lineNum">     347 </span>            :    * @param[in]   CGI_ResultsVector     results</a>
<a name="348"><span class="lineNum">     348 </span>            :    * @param[in]   fileName              file name where results will be reported</a>
<a name="349"><span class="lineNum">     349 </span>            :    */</a>
<a name="350"><span class="lineNum">     350 </span><span class="lineNoCov">          0 :   void outputPhylip(skch::Parameters &amp;parameters,</span></a>
<a name="351"><span class="lineNum">     351 </span>            :       std::unordered_map &lt;std::string, uint64_t&gt; &amp;genomeLengths,</a>
<a name="352"><span class="lineNum">     352 </span>            :       std::vector&lt;cgi::CGI_Results&gt; &amp;CGI_ResultsVector,</a>
<a name="353"><span class="lineNum">     353 </span>            :       std::string &amp;fileName)</a>
<a name="354"><span class="lineNum">     354 </span>            :   {</a>
<a name="355"><span class="lineNum">     355 </span><span class="lineNoCov">          0 :     std::unordered_map &lt;std::string, int&gt; genome2Int;    // name of genome -&gt; integer</span></a>
<a name="356"><span class="lineNum">     356 </span><span class="lineNoCov">          0 :     std::unordered_map &lt;int, std::string&gt; genome2Int_rev; // integer -&gt; name of genome</span></a>
<a name="357"><span class="lineNum">     357 </span>            : </a>
<a name="358"><span class="lineNum">     358 </span>            :     //Assign unique index to the set of query and reference genomes</a>
<a name="359"><span class="lineNum">     359 </span><span class="lineNoCov">          0 :     for(auto &amp;e : parameters.querySequences)</span></a>
<a name="360"><span class="lineNum">     360 </span>            :     {</a>
<a name="361"><span class="lineNum">     361 </span><span class="lineNoCov">          0 :       auto id = genome2Int.size();</span></a>
<a name="362"><span class="lineNum">     362 </span><span class="lineNoCov">          0 :       if( genome2Int.find(e) == genome2Int.end() )</span></a>
<a name="363"><span class="lineNum">     363 </span>            :       {</a>
<a name="364"><span class="lineNum">     364 </span><span class="lineNoCov">          0 :         genome2Int [e] = id;</span></a>
<a name="365"><span class="lineNum">     365 </span><span class="lineNoCov">          0 :         genome2Int_rev [id] = e;</span></a>
<a name="366"><span class="lineNum">     366 </span>            :       }</a>
<a name="367"><span class="lineNum">     367 </span>            :     }</a>
<a name="368"><span class="lineNum">     368 </span>            : </a>
<a name="369"><span class="lineNum">     369 </span><span class="lineNoCov">          0 :     for(auto &amp;e : parameters.refSequences)</span></a>
<a name="370"><span class="lineNum">     370 </span>            :     {</a>
<a name="371"><span class="lineNum">     371 </span><span class="lineNoCov">          0 :       auto id = genome2Int.size();</span></a>
<a name="372"><span class="lineNum">     372 </span><span class="lineNoCov">          0 :       if( genome2Int.find(e) == genome2Int.end() )</span></a>
<a name="373"><span class="lineNum">     373 </span>            :       {</a>
<a name="374"><span class="lineNum">     374 </span><span class="lineNoCov">          0 :         genome2Int [e] = id;</span></a>
<a name="375"><span class="lineNum">     375 </span><span class="lineNoCov">          0 :         genome2Int_rev [id] = e;</span></a>
<a name="376"><span class="lineNum">     376 </span>            :       }</a>
<a name="377"><span class="lineNum">     377 </span>            :     }</a>
<a name="378"><span class="lineNum">     378 </span>            : </a>
<a name="379"><span class="lineNum">     379 </span><span class="lineNoCov">          0 :     int totalGenomes = genome2Int.size();</span></a>
<a name="380"><span class="lineNum">     380 </span>            : </a>
<a name="381"><span class="lineNum">     381 </span>            :     //create a square 2-d matrix</a>
<a name="382"><span class="lineNum">     382 </span><span class="lineNoCov">          0 :     std::vector&lt; std::vector&lt;float&gt; &gt; fastANI_matrix (totalGenomes,  std::vector&lt;float&gt; (totalGenomes, 0.0));</span></a>
<a name="383"><span class="lineNum">     383 </span>            : </a>
<a name="384"><span class="lineNum">     384 </span>            :     //transform FastANI results into 3-tuples</a>
<a name="385"><span class="lineNum">     385 </span><span class="lineNoCov">          0 :     for(auto &amp;e : CGI_ResultsVector)</span></a>
<a name="386"><span class="lineNum">     386 </span>            :     {</a>
<a name="387"><span class="lineNum">     387 </span><span class="lineNoCov">          0 :       std::string qryGenome = parameters.querySequences[e.qryGenomeId];</span></a>
<a name="388"><span class="lineNum">     388 </span><span class="lineNoCov">          0 :       std::string refGenome = parameters.refSequences[e.refGenomeId];</span></a>
<a name="389"><span class="lineNum">     389 </span>            : </a>
<a name="390"><span class="lineNum">     390 </span><span class="lineNoCov">          0 :       assert(genomeLengths.find(qryGenome) != genomeLengths.end());</span></a>
<a name="391"><span class="lineNum">     391 </span><span class="lineNoCov">          0 :       assert(genomeLengths.find(refGenome) != genomeLengths.end());</span></a>
<a name="392"><span class="lineNum">     392 </span>            : </a>
<a name="393"><span class="lineNum">     393 </span><span class="lineNoCov">          0 :       uint64_t queryGenomeLength = genomeLengths[qryGenome];</span></a>
<a name="394"><span class="lineNum">     394 </span><span class="lineNoCov">          0 :       uint64_t refGenomeLength = genomeLengths[refGenome]; </span></a>
<a name="395"><span class="lineNum">     395 </span><span class="lineNoCov">          0 :       uint64_t minGenomeLength = std::min(queryGenomeLength, refGenomeLength);</span></a>
<a name="396"><span class="lineNum">     396 </span><span class="lineNoCov">          0 :       uint64_t sharedLength = e.countSeq * parameters.minReadLength;</span></a>
<a name="397"><span class="lineNum">     397 </span>            : </a>
<a name="398"><span class="lineNum">     398 </span>            :       //Checking if shared genome is above a certain fraction of genome length</a>
<a name="399"><span class="lineNum">     399 </span><span class="lineNoCov">          0 :       if(sharedLength &gt;= minGenomeLength * parameters.minFraction)</span></a>
<a name="400"><span class="lineNum">     400 </span>            :       {</a>
<a name="401"><span class="lineNum">     401 </span><span class="lineNoCov">          0 :         int qGenome = genome2Int [ qryGenome ];</span></a>
<a name="402"><span class="lineNum">     402 </span><span class="lineNoCov">          0 :         int rGenome = genome2Int [ refGenome ];</span></a>
<a name="403"><span class="lineNum">     403 </span>            : </a>
<a name="404"><span class="lineNum">     404 </span><span class="lineNoCov">          0 :         if (qGenome != rGenome)   //ignore if both genomes are same</span></a>
<a name="405"><span class="lineNum">     405 </span>            :         {</a>
<a name="406"><span class="lineNum">     406 </span><span class="lineNoCov">          0 :           if (qGenome &gt; rGenome)</span></a>
<a name="407"><span class="lineNum">     407 </span>            :           {</a>
<a name="408"><span class="lineNum">     408 </span><span class="lineNoCov">          0 :             if (fastANI_matrix[qGenome][rGenome] &gt; 0)</span></a>
<a name="409"><span class="lineNum">     409 </span><span class="lineNoCov">          0 :               fastANI_matrix[qGenome][rGenome] = (fastANI_matrix[qGenome][rGenome] + e.identity)/2;</span></a>
<a name="410"><span class="lineNum">     410 </span>            :             else</a>
<a name="411"><span class="lineNum">     411 </span><span class="lineNoCov">          0 :               fastANI_matrix[qGenome][rGenome] = e.identity;</span></a>
<a name="412"><span class="lineNum">     412 </span>            :           }</a>
<a name="413"><span class="lineNum">     413 </span>            :           else</a>
<a name="414"><span class="lineNum">     414 </span>            :           {</a>
<a name="415"><span class="lineNum">     415 </span><span class="lineNoCov">          0 :             if (fastANI_matrix[rGenome][qGenome] &gt; 0)</span></a>
<a name="416"><span class="lineNum">     416 </span><span class="lineNoCov">          0 :               fastANI_matrix[rGenome][qGenome] = (fastANI_matrix[rGenome][qGenome] + e.identity)/2;</span></a>
<a name="417"><span class="lineNum">     417 </span>            :             else</a>
<a name="418"><span class="lineNum">     418 </span><span class="lineNoCov">          0 :               fastANI_matrix[rGenome][qGenome] = e.identity;</span></a>
<a name="419"><span class="lineNum">     419 </span>            :           }</a>
<a name="420"><span class="lineNum">     420 </span>            :         }</a>
<a name="421"><span class="lineNum">     421 </span>            :       }</a>
<a name="422"><span class="lineNum">     422 </span><span class="lineNoCov">          0 :     }</span></a>
<a name="423"><span class="lineNum">     423 </span>            : </a>
<a name="424"><span class="lineNum">     424 </span><span class="lineNoCov">          0 :     std::ofstream outstrm(fileName + &quot;.matrix&quot;);</span></a>
<a name="425"><span class="lineNum">     425 </span>            : </a>
<a name="426"><span class="lineNum">     426 </span><span class="lineNoCov">          0 :     outstrm &lt;&lt; totalGenomes &lt;&lt; &quot;\n&quot;;</span></a>
<a name="427"><span class="lineNum">     427 </span>            : </a>
<a name="428"><span class="lineNum">     428 </span>            :     //Report matrix</a>
<a name="429"><span class="lineNum">     429 </span><span class="lineNoCov">          0 :     for (int i = 0; i &lt; totalGenomes; i++)</span></a>
<a name="430"><span class="lineNum">     430 </span>            :     {</a>
<a name="431"><span class="lineNum">     431 </span>            :       //output genome name</a>
<a name="432"><span class="lineNum">     432 </span><span class="lineNoCov">          0 :       outstrm &lt;&lt; genome2Int_rev[i];</span></a>
<a name="433"><span class="lineNum">     433 </span>            : </a>
<a name="434"><span class="lineNum">     434 </span><span class="lineNoCov">          0 :       for (int j = 0; j &lt; i; j++)</span></a>
<a name="435"><span class="lineNum">     435 </span>            :       {</a>
<a name="436"><span class="lineNum">     436 </span>            :         //output ani values</a>
<a name="437"><span class="lineNum">     437 </span>            :         //average if computed twice</a>
<a name="438"><span class="lineNum">     438 </span><span class="lineNoCov">          0 :         std::string val = fastANI_matrix[i][j] &gt; 0.0 ? std::to_string (fastANI_matrix[i][j]) : &quot;NA&quot;;</span></a>
<a name="439"><span class="lineNum">     439 </span><span class="lineNoCov">          0 :         outstrm &lt;&lt; &quot;\t&quot; &lt;&lt; val; </span></a>
<a name="440"><span class="lineNum">     440 </span><span class="lineNoCov">          0 :       }</span></a>
<a name="441"><span class="lineNum">     441 </span><span class="lineNoCov">          0 :       outstrm &lt;&lt; &quot;\n&quot;;</span></a>
<a name="442"><span class="lineNum">     442 </span>            :     }</a>
<a name="443"><span class="lineNum">     443 </span>            : </a>
<a name="444"><span class="lineNum">     444 </span><span class="lineNoCov">          0 :     outstrm.close();</span></a>
<a name="445"><span class="lineNum">     445 </span><span class="lineNoCov">          0 :   }</span></a>
<a name="446"><span class="lineNum">     446 </span>            : </a>
<a name="447"><span class="lineNum">     447 </span>            :   /**</a>
<a name="448"><span class="lineNum">     448 </span>            :    * @brief                         generate multiple parameter objects from one</a>
<a name="449"><span class="lineNum">     449 </span>            :    * @details                       purpose it to divide the list of reference genomes</a>
<a name="450"><span class="lineNum">     450 </span>            :    *                                into as many buckets as there are threads </a>
<a name="451"><span class="lineNum">     451 </span>            :    * @param[in]   parameters</a>
<a name="452"><span class="lineNum">     452 </span>            :    * @param[out]  parameters_split</a>
<a name="453"><span class="lineNum">     453 </span>            :    */</a>
<a name="454"><span class="lineNum">     454 </span><span class="lineNoCov">          0 :   void splitReferenceGenomes(skch::Parameters &amp;parameters,</span></a>
<a name="455"><span class="lineNum">     455 </span>            :       std::vector &lt;skch::Parameters&gt; &amp;parameters_split)</a>
<a name="456"><span class="lineNum">     456 </span>            :   {</a>
<a name="457"><span class="lineNum">     457 </span><span class="lineNoCov">          0 :     for (int i = 0; i &lt; parameters.threads; i++)</span></a>
<a name="458"><span class="lineNum">     458 </span>            :     {</a>
<a name="459"><span class="lineNum">     459 </span><span class="lineNoCov">          0 :       parameters_split[i] = parameters;</span></a>
<a name="460"><span class="lineNum">     460 </span>            : </a>
<a name="461"><span class="lineNum">     461 </span>            :       //update the reference genomes list</a>
<a name="462"><span class="lineNum">     462 </span><span class="lineNoCov">          0 :       parameters_split[i].refSequences.clear();</span></a>
<a name="463"><span class="lineNum">     463 </span>            : </a>
<a name="464"><span class="lineNum">     464 </span>            :       //assign ref. genome to threads in round-robin fashion</a>
<a name="465"><span class="lineNum">     465 </span><span class="lineNoCov">          0 :       for (int j = 0; j &lt; parameters.refSequences.size(); j++)</span></a>
<a name="466"><span class="lineNum">     466 </span>            :       {</a>
<a name="467"><span class="lineNum">     467 </span><span class="lineNoCov">          0 :         if (j % parameters.threads == i)</span></a>
<a name="468"><span class="lineNum">     468 </span><span class="lineNoCov">          0 :           parameters_split[i].refSequences.push_back (parameters.refSequences[j]);</span></a>
<a name="469"><span class="lineNum">     469 </span>            :       }</a>
<a name="470"><span class="lineNum">     470 </span>            :     }</a>
<a name="471"><span class="lineNum">     471 </span><span class="lineNoCov">          0 :   }</span></a>
<a name="472"><span class="lineNum">     472 </span>            : </a>
<a name="473"><span class="lineNum">     473 </span>            :   /**</a>
<a name="474"><span class="lineNum">     474 </span>            :    * @brief                             update thread local reference genome ids to global ids</a>
<a name="475"><span class="lineNum">     475 </span>            :    * @param[in/out] CGI_ResultsVector</a>
<a name="476"><span class="lineNum">     476 </span>            :    */</a>
<a name="477"><span class="lineNum">     477 </span><span class="lineNoCov">          0 :   void correctRefGenomeIds (std::vector&lt;cgi::CGI_Results&gt; &amp;CGI_ResultsVector)</span></a>
<a name="478"><span class="lineNum">     478 </span>            :   {</a>
<a name="479"><span class="lineNum">     479 </span><span class="lineNoCov">          0 :     int tid = omp_get_thread_num();</span></a>
<a name="480"><span class="lineNum">     480 </span><span class="lineNoCov">          0 :     int thread_count = omp_get_num_threads(); </span></a>
<a name="481"><span class="lineNum">     481 </span>            :     </a>
<a name="482"><span class="lineNum">     482 </span><span class="lineNoCov">          0 :     for (auto &amp;e : CGI_ResultsVector)</span></a>
<a name="483"><span class="lineNum">     483 </span><span class="lineNoCov">          0 :       e.refGenomeId = e.refGenomeId * thread_count +  tid;</span></a>
<a name="484"><span class="lineNum">     484 </span><span class="lineNoCov">          0 :   }</span></a>
<a name="485"><span class="lineNum">     485 </span>            : }</a>
<a name="486"><span class="lineNum">     486 </span>            : </a>
<a name="487"><span class="lineNum">     487 </span>            : #endif</a>
</pre>
      </td>
    </tr>
  </table>
  <br>

  <table width="100%" border=0 cellspacing=0 cellpadding=0>
    <tr><td class="ruler"><img src="../../../../../../../glass.png" width=3 height=3 alt=""></td></tr>
    <tr><td class="versionInfo">Generated by: <a href="http://ltp.sourceforge.net/coverage/lcov.php" target="_parent">LCOV version 1.14</a></td></tr>
  </table>
  <br>

</body>
</html>
